### PR TITLE
docs(retrieval): complete fine-tuning guide integration

### PR DESCRIPTION
**Pull request by oliverholworthy**
# What does this PR do?

Completes the Fern integration for the retrieval fine-tuning guide and aligns it with the dataset behavior available on current `main`. The change remains documentation-focused.

The main fine-tuning page was carried into Fern by #2391 while this PR was under review. This revision updates that page where current implementation behavior requires it, applies tech-pubs guidance to new or substantively edited retrieval prose, and removes unrelated runtime code, utility scripts, tests, agent guidance, and drive-by copy edits from the original #2306 diff.

# Changelog

- Use `RetrievalDatasetConfig` as the single public configuration for corpus-backed JSON, AutoModel-schema `hf://` sources, and inline JSONL.
- Clarify the runtime one-positive schema, positive cycling, multi-positive expansion, sibling-positive masking, and qrels conversion without duplicating guidance.
- State that qrels are conversion and offline-evaluation inputs, not direct training inputs.
- Clarify that pure inline JSONL has no corpus metadata and must keep `use_dataset_instruction: false`.
- Add retrieval links to the dataset overview, recipe overview, home-page quick links, and guide table.
- Correct the fine-tuning guide for validation group sizing, model support, pure DP/FSDP constraints, distributed masking, LoRA module selection, checkpoint resolution, and trusted local checkpoint loading.
- Document the hard-negative miner as it behaves on current `main`, including cache, overwrite, metadata, score-filtering, and Llama-export limitations.
- Remove commands and guarantees that depend on draft companion PRs.
- Update the bi-encoder and cross-encoder example comments, timeout, public training sources, and source/container command forms needed by the guide.

# Deferred Companion PRs

These follow-ups are intentionally not documented as available in #2306:

- #2905 — preserve and restore bi-encoder export metadata.
- #2906 — harden hard-negative mining cache, scoring, and output behavior.
- #2907 — add HF materialization, positive unrolling, lineage, and mined-output audit utilities.
- Runtime PRD — define the intended behavior when pure inline JSONL sets `use_dataset_instruction: true`: either reject the unsupported configuration with a clear error or return empty instructions. #2306 does not choose or change this behavior.

Unrelated guidance extracted from the original mixed diff:

- #2904 — record reusable, prospective Fern prose-style guidance from the tech-pubs review.
- #2903 — update agent references to the YAML-first CLI.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines.
- [x] Kept runtime behavior unchanged and documented the supported pure-inline configuration.
- [x] Ported the documentation to the current Fern MDX layout.
- [x] Included DCO sign-off in all commits.

# Validation

- `ruff format .` and `ruff check --fix .`.
- `tests/unit_tests/datasets/llm/test_retrieval_dataset.py` — 58 passed.
- `tests/unit_tests/datasets/llm/test_typed_dataset_loader.py` — 41 passed.
- `make -C docs/fern docs-check` — 291 MDX files validated; `fern check` found 0 errors and reported the standard single warning.
- `tools/lint_example_yamls.py --automodel-dir .` — 449 YAML files linted.
- Nightly recipe references are valid; all 15 recipe CI groups generated successfully.
- Both modified retrieval YAMLs resolve through `RecipeConfig` to `RetrievalDatasetConfig`.
- `git diff --check`.

# Additional Information

The guide documents only the three verified source forms: corpus-backed JSON, AutoModel-schema `hf://`, and inline JSONL. It does not claim TSV or CSV support because the inline parser currently accepts JSON and JSONL records only.

The original full PR head remains recoverable at commit `dabb437d7537929391b7451a12c4de3b4da7ccd8`. All current inline review threads are resolved.

GPU and multi-node behavior was reviewed against the implementation and existing tests but was not executed locally.

### CODE REVIEW

File: docs/guides/dataset-overview.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 This page summarizes the datasets supported in NeMo AutoModel for LLM, VLM, and retrieval training and shows how to plug in your own datasets using Python functions or the YAML `_target_` mechanism.
 
-- See also: [LLM datasets](llm/dataset.md), [VLM datasets](vlm/dataset.md), and [Retrieval dataset](llm/retrieval-dataset.md) for deeper, task-specific guides.
+- See also: [LLM datasets](llm/dataset.md), [VLM datasets](vlm/dataset.md), [Retrieval dataset](llm/retrieval-dataset.md), and [Retrieval fine-tuning](llm/retrieval-finetuning.md) for deeper, task-specific guides.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
- Refer to [LLM datasets](llm/dataset.md), [VLM datasets](vlm/dataset.md), [Retrieval dataset](llm/retrieval-dataset.md), and [Retrieval fine-tuning](llm/retrieval-finetuning.md) for deeper, task-specific guidance.
```

File: docs/guides/dataset-overview.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
...
 
 - If a dataset you need is missing, please open a [GitHub issue](https://github.com/NVIDIA-NeMo/Automodel/issues) with a short description and example schema so we can prioritize support.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
- If a dataset you need is missing, open a [GitHub issue](https://github.com/NVIDIA-NeMo/Automodel/issues) with a short description and example schema so we can prioritize support.
```

File: docs/guides/dataset-overview.md
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
    - Supports streaming mode for large datasets. Refer to [Streaming Datasets](#streaming-datasets) below for details.
```

File: docs/guides/dataset-overview.md [hunk 3]
```diff
@@ -62,27 +62,6 @@ dataset:
 ```
 See the detailed guide, [Column-Mapped Text Instruction Dataset](llm/column-mapped-text-instruction-dataset.md), for more information.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to the [Column-Mapped Text Instruction Dataset](llm/column-mapped-text-instruction-dataset.md) for more information.
```

File: docs/guides/dataset-overview.md [hunk 4]
```diff
@@ -237,26 +216,38 @@ dataset:
 See the [Function Calling guide](llm/toolcalling.md) for an end-to-end example with FunctionGemma.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to the [Function Calling guide](llm/toolcalling.md) for an end-to-end example with FunctionGemma.
```

File: docs/guides/dataset-overview.md [hunk 5]
```diff
@@ -237,26 +216,38 @@ dataset:
...
 For a small reasoning-style chat SFT starting point, see [qwen2_5_0p5b_instruct_fineproofs_chat.yaml](../../examples/llm_finetune/qwen/qwen2_5_0p5b_instruct_fineproofs_chat.yaml).
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
For a small reasoning-style chat SFT starting point, refer to [qwen2_5_0p5b_instruct_fineproofs_chat.yaml](../../examples/llm_finetune/qwen/qwen2_5_0p5b_instruct_fineproofs_chat.yaml).
```

File: docs/guides/dataset-overview.md [hunk 6]
```diff
@@ -237,26 +216,38 @@ dataset:
...
+    n_passages: 5
+  collate_fn:
+    _target_: nemo_automodel.components.datasets.llm.BiEncoderCollator
+    q_max_len: 512
+    p_max_len: 512
+  batch_size: 2
+  shuffle: true
 ```
-See the detailed guide, [Retrieval dataset](llm/retrieval-dataset.md), for more information.
+See [Retrieval dataset](llm/retrieval-dataset.md) for schema details and [Retrieval fine-tuning](llm/retrieval-finetuning.md) for the end-to-end workflow.
```
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to [Retrieval dataset](llm/retrieval-dataset.md) for schema details and [Retrieval fine-tuning](llm/retrieval-finetuning.md) for the end-to-end workflow.
```

File: docs/guides/dataset-overview.md
Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to the [pretraining guide](llm/pretraining.md), which uses MegatronPretraining data.
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:315]:
```suggestion
### Decide When to Stream
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:331]:
```suggestion
### Enable Streaming
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:414]:
```suggestion
### Stream a Large Dataset
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:462]:
```suggestion
## VLM Datasets (Vision/Audio and Language)
```

Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Each example follows the conversation schema expected by `apply_chat_template`, as shown below:
```

Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to [Gemma-3n](omni/gemma3-3n.md) and [VLM dataset](vlm/dataset.md) for end-to-end examples.
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:534]:
```suggestion
### Configure a Video Dataloader
```

Comment (jgerh) [inline docs/guides/dataset-overview.md]:
```suggestion
Refer to the [Diffusion Dataset Preparation](diffusion/dataset.md) guide for full preprocessing instructions and configuration details.
```

Comment (jgerh) [inline docs/guides/dataset-overview.md:587]:
```suggestion
### Map Instruction Data with ColumnMappedTextInstructionDataset (LLM)
```

File: docs/guides/llm/retrieval-dataset.md [hunk 7]
```diff
@@ -1,21 +1,64 @@
...
+`datasets.Dataset`. At runtime it transforms each raw record into the training-time schema:
 
 - `question`: query string
 - `doc_text`: list of document texts in the order `[positive, negative_1, negative_2, ...]`
 - `doc_image`: list of images (or empty strings), aligned with `doc_text`
-- `query_instruction` / `passage_instruction`: optional, used when `use_dataset_instruction: true` and the corpus provides instructions via metadata
+- `doc_id`: list of document IDs aligned with `doc_text` for corpus-backed and `hf://` sources. Pure inline JSONL
+  does not provide real document IDs for duplicate-document masking unless you add them in a custom preprocessing path.
+- `query_instruction` / `passage_instruction`: optional, used when `use_dataset_instruction: true` and the corpus
+  provides instructions via metadata
```
Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
  provides instructions through metadata
```

File: docs/guides/llm/retrieval-dataset.md [hunk 8]
```diff
@@ -1,21 +1,64 @@
-# Retrieval Dataset (Embedding Fine-tuning)
+# Retrieval Dataset for Bi-Encoders and Cross-Encoders
 
-NeMo Automodel supports **retrieval model fine-tuning** using a retrieval-style dataset: each training example is a **query** paired with **one positive** document and **one or more negative** documents.
+NeMo AutoModel supports **retrieval model fine-tuning** using a retrieval-style dataset: each training example is a
+**query** paired with **one positive** document and **one or more negative** documents.
 
-This dataset is used by the retrieval recipes (see `examples/retrieval/bi_encoder/` and `examples/retrieval/cross_encoder/`) together with the `BiEncoderCollator`.
+This dataset is used by the retrieval recipes (see `examples/retrieval/bi_encoder/` and
+`examples/retrieval/cross_encoder/`) together with the retrieval collators. For an end-to-end training workflow, see
```
Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
This dataset is used by the retrieval recipes in `examples/retrieval/bi_encoder/` and
`examples/retrieval/cross_encoder/`, together with the retrieval collators. For an end-to-end training workflow, refer to
```

File: docs/guides/llm/retrieval-dataset.md [hunk 9]
```diff
@@ -71,22 +151,25 @@ This is convenient for custom fine-tuning pipelines where the documents are incl
...
   - lists of strings, or
-  - dicts with at least `text` (optionally `image`, `nr_ocr`) for multimodal use cases.
+  - dicts with at least `text`.
+- The current LLM retrieval collators tokenize text only. Do not rely on inline `image` or OCR fields unless you add a
+  custom preprocessing and collator path.
 - If `corpus_id` is not provided, it defaults to `__inline__`.
 - `use_dataset_instruction: true` has no effect for pure inline records (instructions come from corpus metadata).
 :::
 
 ## YAML Usage (Dataset + Collator)
```
Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
## Configure the Dataset and Collator in YAML
```

File: docs/guides/llm/retrieval-dataset.md
Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
JSONL example (one example per line):
```

Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
Training file example (single JSON):
```

File: docs/guides/llm/retrieval-dataset.md [hunk 10]
```diff
@@ -42,19 +85,56 @@ This is the format used by NeMo retriever pipelines where documents live in a se
 
 **Corpus requirements**
```
Comment (jgerh) [inline docs/guides/llm/retrieval-dataset.md]:
```suggestion
Corpus requirements:
```

File: docs/guides/llm/retrieval-finetuning.md [hunk 11]
```diff
@@ -0,0 +1,871 @@
...
+Start with a bi-encoder when you need embeddings for approximate nearest-neighbor search. Add hard-negative mining after
+the first pass if the model mostly sees easy negatives. Train a cross-encoder when a separate retriever already produces
+a small candidate set and you want a stronger reranking stage.
+
+## Quickstart
+
+Before running the examples:
+
+- Use an AutoModel environment with the full GPU training dependencies installed. The NGC container is the safest path
+  for multi-GPU runs; for source checkouts, see [Installation](../installation.md) and run
```
Comment (jgerh) [inline docs/guides/llm/retrieval-finetuning.md]:
```suggestion
  for multi-GPU runs; for source checkouts, refer to [Installation](../installation.md) and run
```

File: docs/guides/llm/retrieval-finetuning.md [hunk 12]
```diff
@@ -0,0 +1,871 @@
...
+  `uv sync --frozen --extra all`.
+- Run the example commands from a source checkout or an NGC/container workspace that contains the repository
+  `examples/` tree. The YAML configs and mining helpers below are repo-relative; if you use an installed package
+  without the repository, copy the referenced config/script files into your own project and update the paths.
+- From a source checkout, use `uv run automodel ...`; from an installed environment that has local copies of the
+  configs, use `automodel ...`.
+- Accept access terms for the configured Hugging Face model and set `HF_TOKEN`, or replace the model path with a model
+  your environment can download. See the support matrix below before swapping model families.
```
Comment (jgerh) [inline docs/guides/llm/retrieval-finetuning.md]:
```suggestion
  your environment can download. Refer to the support matrix below before swapping model families.
```

File: docs/guides/llm/retrieval-finetune.mdx [hunk 13]
```diff
@@ -0,0 +1,871 @@
...
+  --drop-invalid-negatives \
+  --min-negatives 1 \
+  --output /path/to/mined_audited.json
+```
+
+With `--drop-invalid-negatives --output`, the command exits successfully when the cleaned output has no remaining audit
+findings and still satisfies `--min-negatives` if you set it. The audit flags missing or non-finite positive scores, and
+flags and drops negatives whose IDs also appear in the row's `pos_doc`, duplicate negative IDs in the same row, missing
+negative scores, and non-finite negative scores.
+The cleaned output preserves query lineage fields such as `original_question_id`, so unrolled examples remain traceable
```
Comment (jgerh) [inline docs/guides/llm/retrieval-finetune.mdx]:
```suggestion

The cleaned output preserves query lineage fields such as `original_question_id`, so unrolled examples remain traceable
```

File: docs/guides/overview.md
Comment (jgerh) [inline docs/guides/overview.md:8]:
```suggestion
NeMo AutoModel is organized around two key concepts: recipes and components.
```

Comment (jgerh) [inline docs/guides/overview.md:20]:
```suggestion
## Large Language Models (LLM)

This section provides practical recipes and configurations for working with large language models across three core workflows: fine-tuning, pretraining, and knowledge distillation.
```

Comment (jgerh) [inline docs/guides/overview.md:56]:
```suggestion
## Vision Language Models (VLM)

This section provides practical recipes and configurations for working with vision language models, covering fine-tuning and generation workflows for multimodal tasks.
```

File: docs/guides/overview.md [hunk 14]
```diff
@@ -65,22 +75,10 @@ Simple generation script and configurations for VLMs.
...
-
-### Fine-Tuning
-
-End-to-end ASR fine-tuning of `Qwen3-Omni-30B-A3B-Instruct` on any HuggingFace audio dataset, including a thinker-only checkpoint export step for downstream `transformers` / vLLM loading.
-
-- Folder: [examples/audio_finetune/qwen3_omni_asr](https://github.com/NVIDIA-NeMo/Automodel/tree/main/examples/audio_finetune/qwen3_omni_asr)
-- Representative model: Qwen3-Omni-30B-A3B-Instruct
-- How-to guide: [Fine-tune Qwen3-Omni for ASR](audio/qwen3-omni-asr.md)
-
 ## Diffusion Models (Text-to-Image & Text-to-Video)
```
Comment (jgerh) [inline docs/guides/overview.md:84]:
```suggestion
## Diffusion Models (Text-to-Image and Text-to-Video)
```

File: docs/index.md [hunk 15]
```diff
@@ -88,20 +84,19 @@ New models are added regularly. Pick a model below to start fine-tuning, or see
 
 ## Recipes & Guides
```
Comment (jgerh) [inline docs/index.md]:
```suggestion
## Recipes and Guides
```

File: docs/index.md [hunk 16]
```diff
@@ -26,7 +26,7 @@ content_type: index
 
 # NeMo AutoModel Documentation
 
-PyTorch-native training that scales from 1 GPU to thousands with a single config change. Load any Hugging Face model, point at your data, and start training; no checkpoint conversion and no boilerplate.
+PyTorch-native training that scales from 1 GPU to thousands with a single config change. Load any Hugging Face model, point at your data, and start training -- no checkpoint conversion, no boilerplate.
 **Quick links:** [🤗 HF Compatible](guides/huggingface-api-compatibility.md) | [🚀 Performance](performance-summary.md) | [📐 Scalability](about/key-features.md) | [🎯 SFT & PEFT](guides/llm/finetune.md) | [🎨 Diffusion](guides/diffusion/finetune.md) | [👁️ VLM](guides/vlm/gemma4.md)
```
Comment (jgerh) [inline docs/index.md]: consider adding full list of guides

```suggestion
**Quick links:** [🤗 HF Compatible](guides/huggingface-api-compatibility.md) | [🚀 Performance](performance-summary.md) | [📐 Scalability](about/key-features.md) | [🎯 SFT & PEFT](guides/llm/finetune.md) | [🎨 Diffusion](guides/diffusion/finetune.md) | [👁️ VLM](guides/vlm/gemma4.md) | [🔀 dLLM](guides/dllm/finetune.md) | [🔊 Audio](guides/audio/qwen3-omni-asr.md) | [🧩 Omni](guides/omni/gemma3-3n.md) | [⚡ Speculative](guides/speculative/serve_with_sglang.md)
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:7]:
```suggestion
This guide provides step-by-step instructions to reproduce the training pipeline for the `llama-embed-nemotron-8b` model using the [NeMo AutoModel](https://github.com/NVIDIA-NeMo/Automodel) framework.
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:9]:
```suggestion
## Run the Training Pipeline
```

Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:11]:
```suggestion
### Download and Prepare the Dataset
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md [hunk 17]
```diff
@@ -12,20 +12,20 @@ This guide provides step-by-step instructions to reproduce the training pipeline
...
 
-```python
-python examples/retrieval/bi_encoder/llama_embed_nemotron_8b/data_preparation.py \
+```bash
+uv run python examples/retrieval/bi_encoder/llama_embed_nemotron_8b/data_preparation.py \
     --download-path ./embed_nemotron_dataset_v1
 ```
 
-This script will download the dataset and prepare it for training. 
+Run this command from the repository root, or update the relative paths in the YAML. This script will download the dataset and prepare it for training.
```
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md]:
```suggestion
Run this command from the repository root, or update the relative paths in the YAML. This script downloads the dataset and prepares it for training.
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md [hunk 18]
```diff
@@ -12,20 +12,20 @@ This guide provides step-by-step instructions to reproduce the training pipeline
...
+
 
 ### 2. Run Model Finetuning
```
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:22]:
```suggestion
### Fine-Tune the Model
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md [hunk 19]
```diff
@@ -12,20 +12,20 @@ This guide provides step-by-step instructions to reproduce the training pipeline
...
 
 Run the model finetuning with the specified configuration using 8 GPUs:
```
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:24]:
```suggestion
Fine-tune the model on 8 GPUs using the example YAML configuration:
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md [hunk 20]
```diff
@@ -12,20 +12,20 @@ This guide provides step-by-step instructions to reproduce the training pipeline
...
 
 ```bash
-torchrun --nproc-per-node=8 examples/retrieval/bi_encoder/finetune.py \
-    --config examples/retrieval/bi_encoder/llama_embed_nemotron_8b/llama_embed_nemotron_8b.yaml
+uv run automodel examples/retrieval/bi_encoder/llama_embed_nemotron_8b/llama_embed_nemotron_8b.yaml --nproc-per-node 8
 ```
 
 The final model checkpoint in Hugging Face format will be stored in `output/llama_embed_nemotron_8b/epoch_0_step_28614/model/consolidated`
```
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:31]:
```suggestion
The final model checkpoint in Hugging Face format is stored in `output/llama_embed_nemotron_8b/epoch_0_step_28614/model/consolidated`.
```

File: examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md
Comment (jgerh) [inline examples/retrieval/bi_encoder/llama_embed_nemotron_8b/README.md:35]:
```suggestion
If you use this model or training pipeline in your research, cite:
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of .md files and provided some copyedits and suggested text revisions.

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of .md files and provided some copyedits and suggested text revisions.

File: examples/retrieval/data_utils/mine_hard_negatives.py [hunk 21]
```diff
@@ -12,18 +12,26 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# To run this script, use the following command:
-# torchrun --nproc_per_node=8 --master_port=29500 ./examples/retrieval/data_utils/mine_hard_negatives.py \
+# Single-node usage:
+# uv run torchrun --standalone --nproc_per_node=8 ./examples/retrieval/data_utils/mine_hard_negatives.py \
```
Comment (rnyak) [inline examples/retrieval/data_utils/mine_hard_negatives.py]: can we pls clarify where uv run would be more suitable? may be not in a  AM docker image? thanks.

**Comment by oliverholworthy** [comment]
#2306 has been rebuilt on current `main` after the Fern migration.

The accepted tech-pubs edits were ported to the current MDX pages. Runtime changes were extracted into #2905, #2906, and #2907; reusable prose guidance moved to #2904; unrelated CLI agent guidance moved to #2903. The remaining diff is now five documentation pages and two runnable example YAMLs.

File: docs/guides/llm/retrieval-dataset.mdx [hunk 22]
```diff
@@ -43,28 +87,63 @@ This is the format used by NeMo retriever pipelines where documents live in a se
...
+
+The `corpus_id` in `merlin_metadata.json` must match the `corpus_id` in each training record. Relative corpus paths in
+`train.json` are resolved relative to the JSON file.
+
 <Note>
 - `pos_doc` and `neg_doc` can be lists of `{"id": ...}` dicts or raw IDs (they are normalized internally).
-- `cycle_positive_docs` defaults to `false`, which always uses the first positive document. When a training record has multiple positive documents, set `cycle_positive_docs: true` to rotate through them deterministically by epoch order (`epoch % len(pos_doc)`).
-- If you set `use_dataset_instruction: true`, optional fields like `query_instruction` and `passage_instruction` in `merlin_metadata.json` are surfaced to the collator.
-
+- Training uses `pos_doc[0]` as the positive. Additional positives are ignored unless you expand the data before
```
Comment (rnyak) [inline docs/guides/llm/retrieval-dataset.mdx]: @oliverholworthy this info was removed. 
```
`cycle_positive_docs` defaults to `false`, which always uses the first positive document. When a training record has multiple positive documents, set `cycle_positive_docs: true` to rotate through them deterministically by epoch order (`epoch % len(pos_doc)`).
```
can we add it back? I think it is useful to keep it to inform users about such flag.

File: examples/retrieval/bi_encoder/llama3_2_1b.yaml [hunk 23]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 # To run this recipe:
-#   automodel examples/retrieval/bi_encoder/llama3_2_1b.yaml --nproc-per-node 8
+#   uv run automodel examples/retrieval/bi_encoder/llama3_2_1b.yaml --nproc-per-node 8
```
Comment (rnyak) [inline examples/retrieval/bi_encoder/llama3_2_1b.yaml:16]: can we clarify this and add two options, one with `uv` one without. can we explain `uv run ..` can be preferred, if training is not run in a docker env where we already have dependencies? Bcs, `uv` can mess the env up. what do you think?

Comment (oliverholworthy) [reply examples/retrieval/bi_encoder/llama3_2_1b.yaml:16]: Addressed after the rebase. Lines 15–16 show the source-checkout form (`uv run automodel ...`), while lines 17–18 show the container or pre-installed-environment form (`automodel ...`) and explain when to use each.

File: docs/guides/llm/retrieval-dataset.mdx [hunk 24]
```diff
@@ -1,30 +1,79 @@
...
+query-passage pair for reranking.
+
+Each transformed training example uses exactly one positive passage. Corpus-backed JSON and `hf://` sources use the
+first item in `pos_doc` by default. Set `cycle_positive_docs: true` to rotate through multiple positives by epoch.
+Inline JSONL currently always uses the first positive. For multi-positive data, choose a canonical positive, enable
+cycling where supported, expand the record into one example per positive, or add a multi-positive loss/masking strategy.
+If expanded rows for the same query can share a batch, keep distributed in-batch negatives disabled unless you also
+prevent sibling positives from becoming negatives through qrels-aware sampling or masking. Keep the full set of known
+positives in your qrels or corpus metadata for evaluation and false-negative filtering, even when each transformed
+example uses one positive.
```
Comment (rnyak) [inline docs/guides/llm/retrieval-dataset.mdx]: we dont have qrels for training, would it be confusing to mention it? 
"Keep the full set of known
positives in your qrels or corpus metadata for evaluation and false-negative filtering, even when each transformed
example uses one positive."  --> "evaluation" refers to model inference here? or it refers to evaluating that we dont use positives as false negatives? not clear. if former, do we need to mention it?

File: docs/guides/llm/retrieval-dataset.mdx [hunk 25]
```diff
@@ -1,30 +1,79 @@
...
+   `neg_doc`. Use unique `question_id` values within each mining file. Hard-negative mining writes results back by ID.
+3. Put the canonical positive first in `pos_doc`. For corpus-backed JSON and `hf://` sources, you can also set
+   `cycle_positive_docs: true` to rotate through positives by epoch. Expand multi-positive queries into multiple records
+   if every positive must become a supervised positive in the same epoch.
+4. For hard-negative mining, include all known positive document IDs for that query in the row's `pos_doc`. The miner
+   excludes only IDs present in the input row, not an external qrels file.
+5. If you expand one query into multiple positive rows, keep those sibling-positive rows out of the same in-batch-negative
+   training batch or use qrels-aware masking.
+6. Preserve the complete qrels separately for full-corpus evaluation and audit mined negatives against them before
+   reusing the output for training.
```
Comment (rnyak) [inline docs/guides/llm/retrieval-dataset.mdx:69]: several bullets are already mentioned above in `Raw Records and Runtime Schemas`, can we consolidate and shorten them, avoid duplicates?

**Comment by yuhezhang-ai** [comment]
One documentation/API consistency suggestion: could we present `nemo_automodel.components.datasets.llm.retrieval_dataset.RetrievalDatasetConfig` as the single recommended public config for all three documented source types?

`RetrievalDatasetConfig` already handles corpus ID-based `.json`, `hf://`, and inline `.jsonl` inputs internally (as well as inline `.tsv` and `.csv`). The retrieval recipe consumes a typed dataset config through `dataset_config.build()`, while the old `make_*` targets are retained as compatibility mappings. Therefore, telling users to choose between two dataset factories exposes an implementation detail and is inconsistent with the move in this PR to the top-level typed config API.

Suggested wording:

> NeMo AutoModel supports two retrieval record schemas across three source types. Configure all supported sources with `nemo_automodel.components.datasets.llm.retrieval_dataset.RetrievalDatasetConfig`; it selects the appropriate loading path based on the source URI and file type.

I would also use `RetrievalDatasetConfig` in the inline JSONL YAML examples. `InlineRetrievalDatasetConfig` can remain supported without being the recommended path in the user guide. The automatic inline routing is for `.jsonl`, `.tsv`, and `.csv`; an inline `.json` file is not currently detected as inline, but this guide documents inline JSONL.

File: docs/guides/llm/retrieval-dataset.mdx (hunk 25 above)
Comment (oliverholworthy) [reply docs/guides/llm/retrieval-dataset.mdx:69]: Addressed in 8f07cded. The runtime section is now limited to the observable one-positive behavior at lines 31–34. Cycling, expansion, hard-negative mining, sibling-positive masking, and offline evaluation are consolidated into one operational section at lines 51–67.

File: docs/guides/llm/retrieval-dataset.mdx (hunk 24 above)
Comment (oliverholworthy) [reply docs/guides/llm/retrieval-dataset.mdx]: Addressed in 8f07cded. Lines 53–54 now state explicitly that training does not consume qrels as a separate input. Lines 66–67 reserve qrels for offline full-corpus retrieval evaluation and mined-negative auditing, while lines 31–34 keep the runtime contract focused on the single positive emitted per transformed example.

**Comment by oliverholworthy** [comment]
@yuhezhang-ai Implemented in 8f07cded. `RetrievalDatasetConfig` is now the single recommended config across the dataset overview, dataset guide, and fine-tuning guide. The dataset guide documents automatic routing for corpus JSON, `hf://`, and inline JSONL at lines 38–41, and both inline bi-encoder and cross-encoder YAML examples use the public config at lines 203–232. I also added direct bi-/cross-encoder coverage and hardened pure-inline instruction handling. I kept the public claim to the three verified source types: although the router currently recognizes `.tsv` and `.csv` suffixes, the inline parser itself is JSON/JSONL-only today, so the guide does not promise TSV/CSV support.

**Comment by yuhezhang-ai** [comment]
Thanks, @oliverholworthy , and apologies, my earlier suggestion to expose the unified API was premature, as the inline path still had an empty document-ID edge case. I pushed a small fix to preserve the previous inline behavior, added unit tests, and triggered CI. It should be ready to merge after ci.

**Review by yuhezhang-ai — APPROVED** [review]
